src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -1559,33 +1560,32 @@ internal static int IndexOfChar(ref char searchSpace, char value, int length)
             }
             else if (Vector512.IsHardwareAccelerated && length >= Vector512<TValue>.Count)
             {
-                Vector512<TValue> equals, values = Vector512.Create(value);
+                Vector512<TValue> current, values = Vector512.Create(value);
                 ref TValue currentSearchSpace = ref searchSpace;
                 ref TValue oneVectorAwayFromEnd = ref Unsafe.Add(ref searchSpace, length - Vector512<TValue>.Count);
 
                 // Loop until either we've finished all elements or there's less than a vector's-worth remaining.
                 do
                 {
-                    equals = Vector512.LoadUnsafe(ref currentSearchSpace);
+                    current = Vector512.LoadUnsafe(ref currentSearchSpace);
 
-                    if (TNegator.NegateIfNeeded(!Vector512.EqualsAny(values, equals)))
+                    if (TNegator.HasMatch(values, current))
```

Comment:
The root issue for these was that the `V128/V256` path was negating the mask. That is, doing effectively `NegateIfNeeded(Equals(values, current)) != Zero`. Which isn't the same as negating the `EqualsAny`.

Consider if `Equals` returned `00 FF 00 00` and so negating that becomes `FF 00 FF FF` (or vice versa). Neither of these compare equal to `Zero` and so both return `true`. While for `00 00 00 00`, we get `FF FF FF FF` (or vice versa). One of these would compare equal to `Zero` and so would return `false`.

So, we need the concept of `HasMatch` where the `DontNegate` case can do `EqualsAny` (that is `Equals(x, y) != Zero`) and the `Negate` case can do `!EqualsAll` (that is `~Equals(x, y) != Zero`)